Add codegen for implicit `diam` and `area`

JCGoran opened the pull request:
In case `diam` or `area` are not in the PARAMETER or ASSIGNED blocks, the codegen should still succeed to match NOCMODL's behavior.

Commit:
Add codegen for implicit `diam` and `area`

In case `diam` or `area` are not in the PARAMETER or ASSIGNED blocks,
the codegen should still succeed to match NOCMODL's behavior.

Commit:
Remove `diam` and `area` from morphology usecase

Commit:
Merge branch 'master' into jelic/fix_implicit_diam_area

Commit:
Try using fixed mod file

Commit:
Revert "Try using fixed mod file"

This reverts commit 33127b9a8a8d5d20bcb923ce2b2ef9b1b33ec56e.

Commit:
Merge branch 'master' into jelic/fix_implicit_diam_area

Commit:
Fix issue with mismatched mechanisms

JCGoran commented on test/external/CMakeLists.txt:
```diff
@@ -24,8 +24,8 @@ FetchContent_Declare(
-  GIT_REPOSITORY https://github.com/neuronsimulator/reduced_dentate
-  GIT_TAG 58ad3890536ba09e05fd3ee7a178a18270ea8faa
+  GIT_REPOSITORY https://github.com/JCGoran/reduced_dentate
+  GIT_TAG f9d4528adbdb8c74a6e67180e415ba000e97c2c3
```
Okay, this is how the updated version works:

- if `diam` or `area` are used anywhere (i.e. are in some expression), we generate initialization code for them
- if `diam` or `area` are set in a PARAMETER or ASSIGNED blocks, but are _not_ used anywhere else, we emit a warning like `[NMODL] [warning] :: Variable <name> not used anywhere (except possibly VERBATIM block), but declared in a PARAMETER or ASSIGNED block; will generate initialization code for it anyway`, but still generate initialization code for it
- if `diam` or `area` are only used inside of a VERBATIM block, but they are not used anywhere, and they are not declared in ASSIGNED or PARAMETER blocks, we do not generate initialization code for them. This means that the mechanism will probably not work as intended, and may segfault. I do not see a foolproof way to detect `area` or `diam` in a VERBATIM block; I can add a regex for the VERBATIM block, but that may be error-prone (for instance, `\bdiam\b` would erroneously match the string `"contains diam"`, and due to the possibility of macros, parsing C++ is a nightmare).

For the last scenario, I leave it up to the user to be responsible, and either:

- stop using VERBATIM
- actually declare everything that is used in the mod file

nrnhines replied:
That seems good to me.

Commit:
Merge branch 'master' into jelic/fix_implicit_diam_area

Commit:
Add test file for implicit diam and area

Commit:
Only loop over the variables once

Also has better warning handling.

Commit:
Fix issue with capturing structured bindings

It's illegal in C++17. Also, Clang has a bug where this doesn't work
even with `-std=c++20`, so we use an init-capture instead.

JCGoran committed:
Merge branch 'master' into jelic/fix_implicit_diam_area
